### PR TITLE
Add permute_1D_sparse_data Operation for Intel XPU

### PR DESCRIPTION
**Pull request by aagalleg**
Depends on #71 

# Add `permute_1D_sparse_data` Operation for Intel XPU

This PR introduces the `permute_1D_sparse_data` operator to fbgemm-xpu, enabling permutation of sparse data in jagged/1D format on Intel XPU devices.

#### Core Implementation

- **SYCL Kernels** (`permute_1d_sparse_data.cpp/h`): SYCL implementation with three specialized functors for permuting lengths, indices, and weighted data, plus optimized cumsum helpers for offset computation
- **Operator Registration** (`ops_registry.cpp`): Registers the operator with PyTorch's dispatch system using conditional schema registration to avoid conflicts
- **Python API** (`ops.py`): Clean Python wrapper function with type hints supporting optional weights and length sum parameters

#### Infrastructure

- **Build System** (`CMakeLists.txt`): Added SYCL kernel to build configuration

#### Testing

- **Test Suite** (`test_permute_1d_sparse_data.py`):
  - Correctness validation with and without weights
  - Various segment configurations (uniform, mixed lengths, empty segments)
  - Edge cases (empty tensors, single segments, identity permutations)
  - Random permutations with validation
  - CPU-XPU consistency verification

cc: @manuelhsantana, @flezaalv

### CODE REVIEW

**Comment by flezaalv** [comment]
> Depends on #71
> 
> # Add `permute_1D_sparse_data` Operation for Intel XPU
> This PR introduces the `permute_1D_sparse_data` operator to fbgemm-xpu, enabling permutation of sparse data in jagged/1D format on Intel XPU devices.
> 
> ### Changes
> #### Core Implementation
> * **SYCL Kernels** (`permute_1d_sparse_data.cpp/h`): SYCL implementation with three specialized functors for permuting lengths, indices, and weighted data, plus optimized cumsum helpers for offset computation
> * **Operator Registration** (`ops_registry.cpp`): Registers the operator with PyTorch's dispatch system using conditional schema registration to avoid conflicts
> * **Python API** (`ops.py`): Clean Python wrapper function with type hints supporting optional weights and length sum parameters
> 
> #### Infrastructure
> * **Build System** (`CMakeLists.txt`): Added SYCL kernel to build configuration
> 
> #### Testing
> * **Test Suite** (`test_permute_1d_sparse_data.py`):
>   
>   * Correctness validation with and without weights
>   * Various segment configurations (uniform, mixed lengths, empty segments)
>   * Edge cases (empty tensors, single segments, identity permutations)
>   * Random permutations with validation
>   * CPU-XPU consistency verification
> 
> cc: @manuelhsantana, @flezaalv

The test suite has been replaced with the patch `0001-Add-XPU-support-to-fbgemm-tests.patch` for running upstream FBGEMM tests for permute_2d_sparse_data.
Removed ops.py file (no longer necessary).

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 1]
```diff
@@ -87,6 +91,101 @@ index edb0d3615..f825ce27c 100644
...
+-    @unittest.skipIf(*gpu_unavailable)
++    @unittest.skipIf(not xpu_available, "XPU not available")
+     @given(
+         T=st.integers(min_value=1, max_value=20),
+         L=st.integers(min_value=2, max_value=20),
+@@ -232,7 +236,7 @@ class Permute1DSparseFeaturesTest(unittest.TestCase):
+             low=1,
+             high=L,
+             size=(T,),  # 1D tensor with T elements
+-            device=torch.accelerator.current_accelerator(),
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: This should not be needed as `torch.accelerator.current_accelerator()` should return `xpu` as in trivial:
```
$ python -c "import torch; print(torch.xpu.is_available()); print(torch.accelerator.current_accelerator())"
True
xpu
```

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 2]
```diff
@@ -87,6 +91,101 @@ index edb0d3615..f825ce27c 100644
...
++xpu_available: bool = torch.xpu.is_available()
++
+ if open_source:
+     # pyre-ignore[21]
+     from test_utils import gpu_available, gpu_unavailable, on_oss_clang, running_in_oss
+@@ -203,8 +208,7 @@ class PermuteSparseFeaturesTest(unittest.TestCase):
+ 
+ 
+ class Permute1DSparseFeaturesTest(unittest.TestCase):
+-    @unittest.skipIf(*running_in_oss)
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:112]: What is `running_in_oss` and why that's a problem for us?

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 3]
```diff
@@ -87,6 +91,101 @@ index edb0d3615..f825ce27c 100644
...
+-    @unittest.skipIf(*gpu_unavailable)
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:113]: It's better to extend this check rather than hard XPU one, something like:
```
    @unittest.skipIf(*gpu_unavailable and *xpu_unavailable)
```

**Review by dvrogozh — CHANGES_REQUESTED** [review]

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 1 above)
Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: That line will be deleted by the patch and the `device="xpu"` will be added, the result code will be:

```
 lengths = torch.randint(
            low=1,
            high=L,
            size=(T,),  # 1D tensor with T elements
            device="xpu",
        ).type(index_dtype)
```

Comment (dvrogozh) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: I understand what this patch is currently doing and I am arguing that this should not be needed as upstream code should already do what you are trying to achieve - `torch.accelerator.current_accelerator()` should return `xpu`.

Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: got it

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 2 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:112]: FBGEMM has two versions, open source and fbcode (Meta's version for internal use). By default FBGEMM's wheel is compiled with the flag 'open_source' set to True, e.g. 
`
python -c "import fbgemm_gpu; print(getattr(fbgemm_gpu, 'open_source', False))"
True
`
What running_in_oss does is precisely calling the above [getattr](https://github.com/pytorch/FBGEMM/blob/862ae00eff0d0181cd72d863bd5ee9221a778618/fbgemm_gpu/test/test_utils.py#L77):
` 
running_in_oss: Tuple[bool, str] = (
      getattr(fbgemm_gpu, "open_source", False),
      "Test is currently known to fail in OSS mode",
 )
`
And since we are using the wheel and open_source is set to True, by default the test is skipped.

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 1 above)
Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: Reverted, using the original `torch.accelerator.current_accelerator()` instead.

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 3 above)
Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:113]: Done. I extended the condition to:
```
    @unittest.skipIf(
        gpu_unavailable[0] and not torch.xpu.is_available(),
        "GPU and XPU not found",
    )
```

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 1 above)
Comment (dvrogozh) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: Did you verify that it works?

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 4]
```diff
@@ -87,6 +91,33 @@ index edb0d3615..f825ce27c 100644
...
+     # pyre-ignore[21]
+     from test_utils import gpu_available, gpu_unavailable, on_oss_clang, running_in_oss
+@@ -203,8 +206,10 @@ class PermuteSparseFeaturesTest(unittest.TestCase):
+ 
+ 
+ class Permute1DSparseFeaturesTest(unittest.TestCase):
+-    @unittest.skipIf(*running_in_oss)
+-    @unittest.skipIf(*gpu_unavailable)
++    @unittest.skipIf(
++        gpu_unavailable[0] and not torch.xpu.is_available(),
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:115]: Ok for now, but I suspect that going forward we will need to mimic `gpu_unavailable`.

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch [hunk 5]
```diff
@@ -87,6 +91,33 @@ index edb0d3615..f825ce27c 100644
...
++        "GPU and XPU not found",
++    )
+     @given(
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:118]: What this `@given` is for? Why we add this to the test? If that's needed, I suggest to add in-source comment.

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+namespace fbgemm_xpu {
+
+// ============================================================================
+// SYCL Kernel Functors - Operator Implementations
+// ============================================================================
+
+/**
+ * @brief Permute1DLengthsKernel operator implementation
+ */
+template <typename index_t>
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp]: Why we don't parameterize template for permute as in a reference: https://github.com/pytorch/FBGEMM/blob/3bc4f702d9e60c21d6b97150946e2e339625b73a/fbgemm_gpu/src/sparse_ops/sparse_permute_1d.cu#L16?

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @brief Compute exclusive cumsum (prefix sum) for offsets
+ *
+ * output[0] = 0
+ * output[i] = sum(input[0:i]) for i > 0
+ */
+template <typename T>
+at::Tensor exclusive_cumsum_xpu(const at::Tensor& input) {
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:153]: This seems to be defined elsewhere in the cuda reference. Why do we define it here?

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+
+/**
+ * @brief Compute complete cumsum (inclusive + final element)
+ *
+ * output[0] = 0
+ * output[i] = sum(input[0:i]) for i > 0
+ * output[n] = sum(all input)
+ */
+template <typename T>
+at::Tensor complete_cumsum_xpu(const at::Tensor& input) {
```
Comment (dvrogozh) [inline packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:188]: Same as above. This seems to be defined elsewhere in the cuda reference. Why do we define it here?

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 1 above)
Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch]: Yes, I did. The pytest command is already integrated into the `Test fbgemm-xpu` step: https://github.com/intel/torchlib-xpu/actions/runs/30579440544/job/90997602510#step:13:3

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 7 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:153]: exclusive_cumsum_xpu is a version of [asynchronous_exclusive_cumsum_gpu](https://github.com/pytorch/FBGEMM/blob/ea0e623b351bad94ea2208c2dc743b50d07a6490/fbgemm_gpu/src/sparse_ops/sparse_async_cumsum.cu#L61) but relaying torch's cumsum instead of CUDA's cub::DeviceScan::ExclusiveSum.

asynchronous_exclusive_cumsum it's not directly needed by DLRM-v3 in generative recommenders or TorchRec's functions that enable DLRM-v3, because of this we include the function here sincepermute_1D_sparse_data is the only operator that needs it.

cc: @manuelhsantana

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 8 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:188]: @manuelhsantana why does exclusive_cumsum_xpu and complete_cumsum_xpu have the same code? 
Also, the original code of [permute_1D_sparse_data](https://github.com/pytorch/FBGEMM/blob/3bc4f702d9e60c21d6b97150946e2e339625b73a/fbgemm_gpu/src/sparse_ops/sparse_permute_1d.cu#L288) uses asynchronous_complete_cumsum that is an operator that we did implement to enable TorchRec, why wasn't that implementation used?

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 7 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:153]: We can move the code for exclusive_cumsum and complete_cumsum to be in the same files as asyncronous_complete_cumsum like it is done in FBGEMM. For convenience this should be done in PR #73

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 8 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:188]: complete_cumsum_xpu can be changed in PR #73 for asynchronouse_complete_cumsum if they provide the same results. But there is still the mystery of why exclusive_cumsum_xpu and complete_cumsum_xpu have the same code.

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 6 above)
Comment (aagalleg) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp]: Done at commit [ce10815](https://github.com/intel/torchlib-xpu/pull/72/commits/ce108153eb391f29e0f1b5334650b7e32dce9b90)

File: packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch (hunk 5 above)
Comment (flezaalv) [reply packages/fbgemm-xpu/patches/0001-Add-XPU-support-to-fbgemm-tests.patch:118]: As mentioned in the meeting, that line is not modified.

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 7 above)
Comment (dvrogozh) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:153]: Ok, let's do that in PR 73.

**Review by dvrogozh — APPROVED** [review]

**Comment by dvrogozh** [comment]
@aagalleg , this PR requires manual rebase:
```
Unmerged paths:
  (use "git restore --staged <file>..." to unstage)
  (use "git add/rm <file>..." as appropriate to mark resolution)
        both added:      packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h
        deleted by us:   packages/fbgemm-xpu/src/fbgemm_xpu/ops.py
```

And there are other conflicts once you start rebasing.

**Comment by aagalleg** [comment]
> @aagalleg , this PR requires manual rebase:
> 
> ```
> Unmerged paths:
>   (use "git restore --staged <file>..." to unstage)
>   (use "git add/rm <file>..." as appropriate to mark resolution)
>         both added:      packages/fbgemm-xpu/src/fbgemm_xpu/fbgemm_utils/torch_library.h
>         deleted by us:   packages/fbgemm-xpu/src/fbgemm_xpu/ops.py
> ```
> 
> And there are other conflicts once you start rebasing.

Done.

File: packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp (hunk 8 above)
Comment (dvrogozh) [reply packages/fbgemm-xpu/src/fbgemm_xpu/sycl_kernels/permute_1d_sparse_data.cpp:188]: Ok, I am fine to defer that to PR 73. Bottom line is that we better keep sources aligned with original implementation as close as possible for the easier support.
